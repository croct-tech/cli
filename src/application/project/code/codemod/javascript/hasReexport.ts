import * as t from '@babel/types';
import traverse from '@babel/traverse';
import {parse} from '@/application/project/code/codemod/parser';

export type ExportMatcher = {
    moduleName: string | RegExp,
    importName?: string | RegExp,
};

export function hasReexport(source: string | t.File, matcher: ExportMatcher): boolean {
    const ast = typeof source === 'string'
        ? parse(source, ['jsx', 'typescript'])
        : source;

    let found = false;

    traverse(ast, {
        // export * from 'something'
        ExportAllDeclaration: path => {
            const {node} = path;

            if (!t.isStringLiteral(node.source) || !matches(node.source, matcher.moduleName)) {
                return path.skip();
            }

            found = true;

            return path.skip();
        },

        // export {something} from 'something'
        // export {something as somethingElse} from 'something'
        ExportNamedDeclaration: path => {
            const {node} = path;

            if (!t.isStringLiteral(node.source) || !matches(node.source, matcher.moduleName)) {
                return path.skip();
            }

            if (matcher.importName === undefined) {
                found = true;

                return path.skip();
            }

            for (const specifier of node.specifiers) {
                if (t.isExportSpecifier(specifier) && matches(specifier.local, matcher.importName)) {
                    found = true;

                    return path.skip();
                }
            }

            return path.skip();
        },
    });

    return found;
}

function matches(value: t.Identifier|t.StringLiteral, matcher: string | RegExp): boolean {
    const name = t.isIdentifier(value) ? value.name : value.value;

    if (typeof matcher === 'string') {
        return name === matcher;
    }

    return matcher.test(name);
}
