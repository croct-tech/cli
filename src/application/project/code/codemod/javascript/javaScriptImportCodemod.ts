import {File} from '@babel/types';
import {Codemod, ResultCode} from '@/application/project/code/codemod/codemod';
import {ImportTransformer} from '@/application/project/code/codemod/globImportCodemod';
import {transformImports} from '@/application/project/code/codemod/javascript/utils/transformImports';

export type JavaScriptImportOptions = {
    transformer: ImportTransformer,
};

export class JavaScriptImportCodemod implements Codemod<File, JavaScriptImportOptions> {
    public async apply(file: File, options: JavaScriptImportOptions): Promise<ResultCode<File>> {
        const modified = await transformImports(file, options.transformer);

        return {
            result: file,
            modified: modified,
        };
    }
}
