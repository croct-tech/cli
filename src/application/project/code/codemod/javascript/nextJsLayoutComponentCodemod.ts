import * as t from '@babel/types';
import {Codemod, ResultCode} from '@/application/project/code/codemod/codemod';
import {addImport} from '@/application/project/code/codemod/javascript/utils/addImport';

export type LayoutComponentOptions = {
    typescript?: boolean,
};

export type LayoutComponentConfiguration = {
    provider: {
        component: string,
        module: string,
    },
};

export class NextJsLayoutComponentCodemod implements Codemod<t.File, LayoutComponentOptions> {
    private configuration: LayoutComponentConfiguration;

    public constructor(configuration: LayoutComponentConfiguration) {
        this.configuration = configuration;
    }

    public apply(input: t.File, options: LayoutComponentOptions = {}): Promise<ResultCode<t.File>> {
        const isTypescript = options.typescript ?? false;
        const {body} = input.program;

        const providerImport = addImport(input, {
            type: 'value',
            moduleName: this.configuration.provider.module,
            importName: this.configuration.provider.component,
        });

        const propsImport = isTypescript
            ? addImport(input, {
                type: 'type',
                moduleName: 'react',
                importName: 'PropsWithChildren',
            })
            : null;

        const reactNodeImport = isTypescript
            ? addImport(input, {
                type: 'type',
                moduleName: 'react',
                importName: 'ReactNode',
            })
            : null;

        const props = t.objectPattern([
            t.objectProperty(
                t.identifier('children'),
                t.identifier('children'),
            ),
        ]);

        if (propsImport !== null) {
            props.typeAnnotation = t.tsTypeAnnotation(
                t.tsTypeReference(t.identifier(propsImport.localName)),
            );
        }

        const functionDeclaration = t.functionDeclaration(
            t.identifier('RootLayout'),
            [props],
            t.blockStatement([
                t.returnStatement(
                    t.parenthesizedExpression(
                        t.jsxElement(
                            t.jsxOpeningElement(
                                t.jsxIdentifier('html'),
                                [t.jsxAttribute(t.jsxIdentifier('lang'), t.stringLiteral('en'))],
                            ),
                            t.jsxClosingElement(t.jsxIdentifier('html')),
                            [
                                t.jsxText('\n'),
                                t.jsxElement(
                                    t.jsxOpeningElement(t.jsxIdentifier('body'), []),
                                    t.jsxClosingElement(t.jsxIdentifier('body')),
                                    [
                                        t.jsxText('\n'),
                                        t.jsxElement(
                                            t.jsxOpeningElement(t.jsxIdentifier(providerImport.localName), []),
                                            t.jsxClosingElement(t.jsxIdentifier(providerImport.localName)),
                                            [
                                                t.jsxText('\n'),
                                                t.jsxExpressionContainer(t.identifier('children')),
                                                t.jsxText('\n'),
                                            ],
                                        ),
                                        t.jsxText('\n'),
                                    ],
                                ),
                                t.jsxText('\n'),
                            ],
                        ),
                    ),
                ),
            ]),
        );

        if (reactNodeImport !== null) {
            functionDeclaration.returnType = t.tsTypeAnnotation(
                t.tsTypeReference(t.identifier(reactNodeImport.localName)),
            );
        }

        body.push(t.exportDefaultDeclaration(functionDeclaration));

        return Promise.resolve({
            modified: true,
            result: input,
        });
    }
}
