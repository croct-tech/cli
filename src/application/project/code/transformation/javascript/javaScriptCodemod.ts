import {File} from '@babel/types';
import {parse, print} from 'recast';
import {parse as babelParser} from 'recast/parsers/babel-ts.js';
import {Codemod, CodemodOptions, ResultCode} from '@/application/project/code/transformation/codemod';
import {Language} from '@/application/project/code/transformation/javascript/utils/parse';

export type Configuration<O extends CodemodOptions> = {
    codemod: Codemod<File, O>,
    languages: Language[],
};

export class JavaScriptCodemod<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<File, O>;

    public constructor(configuration: Configuration<O>) {
        this.codemod = configuration.codemod;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const ast = parse(input, {
            parser: {
                parse: babelParser,
            },
        });

        const result = await this.codemod.apply(ast, options);

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: print(result.result, {
                reuseWhitespace: false,
            }).code,
        };
    }
}
