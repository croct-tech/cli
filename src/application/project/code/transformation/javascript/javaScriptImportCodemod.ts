import {File} from '@babel/types';
import {Codemod, ResultCode} from '@/application/project/code/transformation/codemod';
import {ImportTransformer} from '@/application/project/code/transformation/globImportCodemod';
import {transformImports} from '@/application/project/code/transformation/javascript/utils/transformImports';

export type JavaScriptImportOptions = {
    transformer: ImportTransformer,
};

export class JavaScriptImportCodemod implements Codemod<File, JavaScriptImportOptions> {
    public async apply(file: File, options: JavaScriptImportOptions): Promise<ResultCode<File>> {
        const modified = await transformImports(file, options.transformer);

        return {
            result: file,
            modified: modified,
        };
    }
}
