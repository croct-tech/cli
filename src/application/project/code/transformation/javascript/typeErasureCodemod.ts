import {File, Comment, noop} from '@babel/types';
import type {VisitNodeObject, Node} from '@babel/traverse';
import {transformFromAstAsync, createConfigItem} from '@babel/core';
import bts from '@babel/plugin-transform-typescript';
import bsd from '@babel/plugin-syntax-decorators';
import {Codemod, ResultCode} from '@/application/project/code/transformation/codemod';
import {isTypescript} from '@/application/project/code/transformation/javascript/utils/isTypescript';

export type TypeErasureCodemodOptions = Record<string, never>;

export class TypeErasureCodemod implements Codemod<File, TypeErasureCodemodOptions> {
    public async apply(file: File): Promise<ResultCode<File>> {
        if (!isTypescript(file)) {
            return {
                modified: false,
                result: file,
            };
        }

        const babelTsTransform = createConfigItem([bts, {onlyRemoveTypeImports: true}]);
        const babelDecoratorSyntax = createConfigItem([bsd, {legacy: true}]);

        // When removing TS-specific constructs (e.g., interfaces),
        // ensure that any comments associated with those constructs are also removed.
        // Otherwise, leftover comments might reference nonexistent code.
        const removeComments: VisitNodeObject<unknown, Node> = {
            enter: function visit(nodePath) {
                const leadingComments = nodePath.node.leadingComments ?? nodePath.node.innerComments ?? null;

                if (leadingComments === null) {
                    return;
                }

                const preservedComments: Comment[] = [];

                for (const comment of leadingComments) {
                    const tokens = comment.loc?.tokens ?? [];
                    const tokenIndex = tokens.findIndex(
                        candidate => (
                            candidate.loc?.start === comment.loc?.start
                            && candidate.loc?.end === comment.loc?.end
                        ),
                    );

                    if (tokenIndex >= 0 && tokenIndex < tokens.length - 1) {
                        const commentToken = tokens[tokenIndex];
                        const nextToken = tokens.find(
                            (candidate, index) => (
                                index > tokenIndex
                                && !['CommentLine', 'CommentBlock'].includes(candidate.type)
                            ),
                        );

                        if (nextToken === undefined) {
                            continue;
                        }

                        const difference = nextToken.loc.start.line - commentToken.loc.end.line;

                        if (difference > 1) {
                            preservedComments.push(comment);
                        }
                    }
                }

                if (preservedComments.length > 0) {
                    const node = noop();

                    node.comments = preservedComments;

                    nodePath.insertBefore(node);
                }
            },
        };

        const result = await transformFromAstAsync(file, undefined, {
            plugins: [
                {
                    name: 'comment-remover',
                    visitor: {
                        Program: removeComments,
                        TSTypeAliasDeclaration: removeComments,
                        TSInterfaceDeclaration: removeComments,
                        TSDeclareFunction: removeComments,
                        TSDeclareMethod: removeComments,
                        TSImportType: removeComments,
                        TSModuleDeclaration: removeComments,
                    },
                },
                babelTsTransform,
                babelDecoratorSyntax,
            ],
            ast: true,
            configFile: false,
        });

        return {
            result: result?.ast ?? file,
            modified: (result?.ast ?? null) !== null,
        };
    }
}
