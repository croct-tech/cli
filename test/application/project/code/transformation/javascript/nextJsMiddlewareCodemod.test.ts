import {resolve} from 'path';
import {
    MiddlewareConfiguration,
    NextJsMiddlewareCodemod,
} from '@/application/project/code/transformation/javascript/nextJsMiddlewareCodemod';
import {loadFixtures} from '../fixtures';
import {JavaScriptCodemod} from '@/application/project/code/transformation/javascript/javaScriptCodemod';

describe('NextJsMiddlewareCodemod', () => {
    const defaultOptions: MiddlewareConfiguration = {
        matcherPattern: '/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)',
        import: {
            module: '@croct/plug-next/middleware',
            middlewareFactoryName: 'withCroct',
            middlewareName: 'middleware',
        },
    };

    const scenarios = loadFixtures<MiddlewareConfiguration>(
        resolve(__dirname, '../fixtures/nextjs-middleware'),
        defaultOptions,
        {
            'matcherAlias.ts': {
                import: {
                    ...defaultOptions.import,
                },
            },
        },
    );

    it.each(scenarios)('should correctly transform $name', async ({name, fixture, options}) => {
        const transformer = new JavaScriptCodemod({
            languages: ['typescript'],
            codemod: new NextJsMiddlewareCodemod(options),
        });

        const output = await transformer.apply(fixture);

        expect(output.result).toMatchSnapshot(name);
    });
});
