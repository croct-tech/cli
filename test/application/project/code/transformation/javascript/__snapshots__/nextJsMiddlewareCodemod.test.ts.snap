// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextJsMiddlewareCodemod should correctly transform configAfterDefaultExport.ts: configAfterDefaultExport.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export const config = {
    matcher: [
        '/((?!api|_next/static|_next/image|favicon.ico).*)',
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request: NextRequest) {
        console.log(request.url);
    }
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform configAfterMiddlewareWithReference.ts: configAfterMiddlewareWithReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

// pattern
const PATTERN = '/((?!api|_next/static|_next/image|favicon.ico).*)';

// matcher
class Matcher {
    static readonly PATTERN = PATTERN;
}

// getMatcher
function getMatcher() {
    return Matcher.PATTERN;
}

// currentMatcher
const currentMatcher = getMatcher();

// currentConfig
const currentConfig = {
    matcher: [
        ...(Array.isArray(currentMatcher) ? currentMatcher : [currentMatcher]),
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ],
}

export const middleware = withCroct({
    matcher: currentConfig.matcher,

    next: function(request: NextRequest) {
        console.log(request.url);
    }
});

function unrelated() {
}

// Non-global scope
{
    // Different scope, should not be considered
    const PATTERN = null;

    function getMatcher() {
    }

    class Matcher {
    }
}

// config
export const config = currentConfig;
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform configAfterNamedExport.ts: configAfterNamedExport.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export const config = {
    matcher: [
        '/((?!api|_next/static|_next/image|favicon.ico).*)',
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ],
}

export default withCroct({
    matcher: config.matcher,

    next: function middleware(request: NextRequest) {
        console.log(request.url);
    }
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform configInvalidReference.ts: configInvalidReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export default withCroct(function middleware(request: NextRequest) {
    console.log(request.url);
});

export const config = bar;
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform configWithArrayMatcher.ts: configWithArrayMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
import type { NextRequest } from 'next/server'

export const config = {
    matcher: [
        '/((?!api|_next/static|_next/image|favicon.ico).*)',
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request: NextRequest) {
        console.log(request.url);
    }
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform configWithIndirectVariableReference.ts: configWithIndirectVariableReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const regex = '/((?!api|_next/static|_next/image|favicon.ico).*)';

const configValue = {
    matcher: [
        ...(Array.isArray(regex) ? regex : [regex]),
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ],
}

const indirectReference = configValue;
export const config = indirectReference;

export const middleware = withCroct({
    matcher: configValue.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform configWithStringMatcher.ts: configWithStringMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const config = {
    matcher: [
        '/((?!api|_next/static|_next/image|favicon.ico).*)',
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform configWithVariableMatcher.ts: configWithVariableMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const regex = '/((?!api|_next/static|_next/image|favicon.ico).*)';

export const config = {
    matcher: [
        ...(Array.isArray(regex) ? regex : [regex]),
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ],
}

export const middleware = withCroct({
    matcher: config.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform configWithVariableReference.ts: configWithVariableReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const regex = '/((?!api|_next/static|_next/image|favicon.ico).*)';

const configValue = {
    matcher: [
        ...(Array.isArray(regex) ? regex : [regex]),
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ],
}

export const config = configValue;

export const middleware = withCroct({
    matcher: configValue.matcher,

    next: function(request) {
        console.log(request.url);
    }
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform configWithoutMatcher.ts: configWithoutMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const config = {
}

export const middleware = withCroct(function(request) {
    console.log(request.url);
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform defaultExportAnonymousFunction.ts: defaultExportAnonymousFunction.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(function() {
    console.log('middleware');
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform defaultExportArrowFunction.ts: defaultExportArrowFunction.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(() => {
    console.log('middleware');
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform defaultExportArrowFunctionReference.ts: defaultExportArrowFunctionReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
const anything = withCroct((request) => console.log(request.url));
export default anything;
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform defaultExportArrowFunctionWithBodyReference.ts: defaultExportArrowFunctionWithBodyReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

const anything = withCroct((request) => {
    console.log(request.url);
})

export default anything;
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform defaultExportClass.ts: defaultExportClass.ts 1`] = `
"export default class Middleware {
    // invalid
}

export { middleware } from "@croct/plug-next/middleware";

export const config = {
    matcher: "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
};
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform defaultExportFunctionDeclaration.ts: defaultExportFunctionDeclaration.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(function anything(request) {
    console.log(request.url);
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform defaultExportFunctionDeclarationReference.ts: defaultExportFunctionDeclarationReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

const anything = withCroct(function(request) {
    console.log(request.url);
});

export default anything;
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform defaultExportFunctionExpressionReference.ts: defaultExportFunctionExpressionReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

const anything = withCroct(function (request) {
    console.log(request.url);
})

export default anything;
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform defaultExportIndirectReference.ts: defaultExportIndirectReference.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

const anything = function (request) {
    console.log(request.url);
}

const something = withCroct(anything);
export default something;
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform empty.ts: empty.ts 1`] = `
"export { middleware } from "@croct/plug-next/middleware";

export const config = {
  matcher: "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
};"
`;

exports[`NextJsMiddlewareCodemod should correctly transform existingAliasedHofCall.ts: existingAliasedHofCall.ts 1`] = `
"import { withCroct as croctMiddleware } from "@croct/plug-next/middleware";

export default croctMiddleware(function () {
    console.log('middleware');
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform existingAliasedMiddlewareCall.ts: existingAliasedMiddlewareCall.ts 1`] = `
"import { middleware } from "@croct/plug-next/middleware";

export default middleware(function () {
    console.log('middleware');
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform existingConfig.ts: existingConfig.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const config = {
    matcher: [
        '/((?!api|_next/static|_next/image|favicon.ico).*)',
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ],
}

export default withCroct({
    matcher: config.matcher
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform existingConfigArrayMatcher.ts: existingConfigArrayMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const config = {
    matcher: ['/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)'],
}

export default withCroct({
    matcher: config.matcher
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform existingConfigMatcher.ts: existingConfigMatcher.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const config = {
    matcher: '/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)',
}

export default withCroct({
    matcher: config.matcher
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform existingHofCall.ts: existingHofCall.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export default withCroct(function () {
    console.log('middleware');
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform existingImport.ts: existingImport.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const config = {
    matcher: [
        '.*',
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ]
};

export const middleware = withCroct({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform existingImportAliased.ts: existingImportAliased.ts 1`] = `
"import { withCroct as croctMiddleware } from "@croct/plug-next/middleware";

export const config = {
    matcher: [
        '.*',
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ]
};

export const middleware = croctMiddleware({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform existingMiddlewareCall.ts: existingMiddlewareCall.ts 1`] = `
"import { middleware as croctMiddleware } from "@croct/plug-next/middleware";

export default croctMiddleware(function () {
    console.log('middleware');
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform existingMiddlewareReexport.ts: existingMiddlewareReexport.ts 1`] = `
"export { middleware } from "@croct/plug-next/middleware";

export const config = {
  matcher: "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
};
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform matcherAlias.ts: matcherAlias.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const config = {
    matcher: [
        '.*',
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ]
};

export const middleware = withCroct({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform namedExportArrowFunction.ts: namedExportArrowFunction.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";
export const middleware = withCroct(() => console.log('middleware'));
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform namedExportArrowFunctionWithBody.ts: namedExportArrowFunctionWithBody.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const middleware = withCroct(() => {
    console.log('middleware');
})
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform namedExportFunctionDeclaration.ts: namedExportFunctionDeclaration.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const middleware = withCroct(function() {
    console.log('middleware');
});
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform namedExportFunctionExpression.ts: namedExportFunctionExpression.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export const middleware = withCroct(function() {
    console.log('middleware');
})
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform namedSpecifiedExport.ts: namedSpecifiedExport.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

const config = {
    matcher: [
        '.*',
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ]
};

const middleware = withCroct({
    matcher: config.matcher,

    next: function() {
        console.log('middleware');
    }
})

export { middleware, config };
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform specifiedExportWithAliases.ts: specifiedExportWithAliases.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

function unrelated() {
}

const _config = {
    matcher: [
        '.*',
        "/((?!api|_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*)"
    ]
};

const _middlewareFn = withCroct({
    matcher: _config.matcher,

    next: function() {
        console.log('middleware');
    }
})

export {
    _middlewareFn as default,
    _config as config
};
"
`;

exports[`NextJsMiddlewareCodemod should correctly transform unrelatedExports.ts: unrelatedExports.ts 1`] = `
"import { withCroct } from "@croct/plug-next/middleware";

export function foo() {
}

export default withCroct(() => {

});

export function middleware() {
}
"
`;
